### PR TITLE
PLT-6484 Add /leave command to leave a channel

### PR DESCRIPTION
**Pull request by dmeza**
#### Summary
Type command `/leave` and be able to leave the current channel and redirected to `Reception (town square)`. Have the command hint display. 

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6484

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by dmeza** [comment]
@cpanato sounds good. Where are the current commands for `apiv4` like `join`, so that I can implement in a similar way.

**Comment by jasonblais** [comment]
@dmeza All APIv4 endpoints are documented here: https://api.mattermost.com/v4

I don't think we have an endpoint for joining a channel, but maybe that can be helpful.

If not, let us know

**Comment by jasonblais** [comment]
I'll add a Work in Progress label for now

If you have any questions @dmeza, you're welcome to post in our [APIv4 channel in pre-release.mattermost.com](https://pre-release.mattermost.com/core/channels/apiv4).

Here's also a list of PRs that adds APIv4 endpoints, in case helpful: https://github.com/mattermost/platform/pulls?q=is%3Apr+apiv4+is%3Aclosed

**Comment by dmeza** [comment]
@cpanato: I got confused by the comment that this should be implemented for `apiv4`. I was trying to look at what needed to be implemented and running in circles. The documentation pointed by @jasonblais does not have information for `command/execute`. 
I tested with curl commands the same as for `apiv3` and it works for `apiv4` as it is: 

`curl -i -d '{"login_id":"someone@nowhere.com","password":"thisisabadpassword"}' http://localhost:8065/api/v4/users/login`

`curl -i -d '{"command": "/leave ", "channel_id": "pxzkgbhsgffnjguetzocxknbur"}' -H 'Authorization: Bearer kmo78755fp8mznygyc5zxexxty' http://localhost:8065/api/v4/commands/execute`

If my tests are correct can this be approved?

File: i18n/en.json [hunk 1]
```diff
@@ -580,6 +580,30 @@
     "translation": "Joined channel."
   },
   {
+    "id": "api.command_leave.desc",
+    "translation": "Leave current channel"
```
Comment (jasonblais) [inline i18n/en.json]: Propose `Leave the current channel.`

File: i18n/en.json [hunk 2]
```diff
@@ -580,6 +580,30 @@
...
+    "id": "api.command_leave.fail.app_error",
+    "translation": "An error occurred while leaving the channel."
+  },
+  {
+    "id": "api.command_leave.list.app_error",
+    "translation": "An error occurred while listing channels."
+  },
+  {
+    "id": "api.command_leave.missing.app_error",
+    "translation": "We couldn't find the channel"
```
Comment (jasonblais) [inline i18n/en.json]: missing period at the end

File: i18n/en.json [hunk 3]
```diff
@@ -580,6 +580,30 @@
...
+  },
+  {
+    "id": "api.command_leave.name",
+    "translation": "leave"
+  },
+  {
+    "id": "api.command_leave.success",
+    "translation": "Left channel."
```
Comment (jasonblais) [inline i18n/en.json]: Propose `Left the channel.`

**Comment by jasonblais** [comment]
@cpanato also feel free to comment on David's note above about it working for both v3 and v4.

**Comment by cpanato** [comment]
@dmeza I'm sorry, I read fast and thought this was new REST endpoint. But it is and new command. That's totally fine and good.

Now, I have one question maybe to @jasonblais or @jwilander: the command tests are in the `API` folder, should we move that to the `app` folder? I got confused since those commands are not part specific for the `API`. I can move to `app` folder or to another one.

File: api/command_leave_test.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
```
Comment (cpanato) [inline api/command_leave_test.go]: update here to follow the standard

`// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.`

File: app/command_leave.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
```
Comment (cpanato) [inline app/command_leave.go]: same here

**Comment by jasonblais** [comment]
@dmeza let me know if you have any questions about my note above?

Would love to accept this for Mattermost 3.10, there's still a few more days before code complete.

**Comment by dmeza** [comment]
@jasonblais: I commited some changes:
1. /leave command in Town Square. I used an existing internationalized message: "api.channel.remove.default.app_error" that displays like this: screen shot 2017-06-13 at 12 03 11 pm
2. made the small changes that you and @cpanato requested.
3. /leave command now performs the same actions as clicking "x" for Direct and Group message channels: plt-6484_leavedirectorgroupchannel
4. For private channels, I think we should refactor the current duplicate code for the leave private channel modal into a new component and have it show/hide based on an action. This could be a single modal as part of the layout and not in multiple places.
Right now the exact same code is in: `components\navbar.jsx#createLeaveChannelModal()` and `components\channel_header.jsx#createLeaveChannelModal()`. I would be adding two more places with the same code.
If this refactoring sounds good, I will need some guidance from someone with more experience on the code base, so that we don't go back and forth as much.

**Comment by jasonblais** [comment]
Thanks @dmeza! I'll give this PR a second try soon.

@jwilander any thoughts on the David's note below?

> For private channels, I think we should refactor the current duplicate code for the leave private channel modal into a new component and have it show/hide based on an action. This could be a single modal as part of the layout and not in multiple places.
>Right now the exact same code is in: components\navbar.jsx#createLeaveChannelModal() and components\channel_header.jsx#createLeaveChannelModal(). I would be adding two more places with the same code.
>If this refactoring sounds good, I will need some guidance from someone with more experience on the code base, so that we don't go back and forth as much.

**Comment by jwilander** [comment]
Sounds good to me 👍

**Comment by jasonblais** [comment]
Thanks @dmeza! 

For 1, that sounds good, keeps the code simpler. We should update the UI text slightly so that it's more consistent with our other system messages. I will make the change myself in a separate PR.

2 & 3 look great.

For 4, would you like to post in the [Developers channel on pre-release](https://pre-release.mattermost.com/core/channels/developers) and either someone from the community or our team will help? Might be easier. Thanks for offering to do this.

The PR is almost there, will be great to add it.

**Comment by jasonblais** [comment]
@dmeza Actually, for 1), propose we use `api.channel.leave.default.app_error` instead of `api.channel.remove.default.app_error`.

```
Cannot leave the default channel {{.Channel}}
```

**Comment by dmeza** [comment]
@jasonblais @jwilander added the change for 1) error message, also the changes for 4) `/leave` private channels, refactored the code to use a new component `webapp/components/modals/leave_private_channel_modal.jsx` and that way there's no repeated code. Now we have logic for '/leave' command on private channels to show a confirmation dialog.

**Comment by jasonblais** [comment]
Thanks @dmeza! And really sorry for the delay, I missed your updates until a couple of days ago when you messaged me.

1 - When I use the `/leave` command in Town Square, it looks like the `api.channel.remove.default.app_error` message is posted instead of `api.channel.leave.default.app`. Can you help take a look?

2 - 4 - looks great.

I found two other corner cases which I'll queue for our UX meeting to review. No action yet.

5 - Whether you should be able to leave a private channel if you're the last member (right now we don't allow that as we want to avoid orphaned channels, but we've been thinking of changing this)

https://pre-release.mattermost.com/core/pl/1gs71sym1bdpif7gfr5k6fz84w

6 - Whether `/leave` should work in reply threads on the right-hand sidebar, since it's not obvious which channel you should leave (the one in the center or on the right)

https://pre-release.mattermost.com/core/pl/64uyaq5ic7bi7rgb8kswjm8qby

---

7 - Looks like the command descriptions in our autocomplete do not end in a period. Would it be okay if I removed the period from `api.command_leave.desc`? It was my mistake as I gave you the UI string so I'm happy to correct it.

image

**Comment by jasonblais** [comment]
Hey @dmeza, just following up on my last post after a discussion with others. This PR is really really close, sorry for the additional change in 6) below.

1 - When I use the `/leave` command in Town Square, it looks like the `api.channel.remove.default.app_error` message is posted instead of `api.channel.leave.default.app`. Can you help take a look?

2 - 4 look great

5 - No changes needed for this.

6 - Can we disable `/leave` command in reply threads on the right-hand sidebar, since it's not obvious which channel you should leave (the one in the center or on the right).

When someone tries to use the `/leave` command in the reply thread, propose 
 - the user doesn't leave a channel
 - ephemeral message is posted `/leave is not supported in reply threads. Use it in the center channel instead.`

---

7 - Looks like the command descriptions in our autocomplete do not end in a period. Would it be okay if I removed the period from `api.command_leave.desc`? It was my mistake as I gave you the UI string so I'm happy to correct it.

image

**Comment by dmeza** [comment]
Hey @jasonblais finished the changes you requested 1 and 7 (small changes). For 6 added the ephemeral message. Attached a gif to show the logic:
plt-6484_errorleavereplyrhs

With these all changes would be ready. Let me know. cc @jwilander

**Comment by jasonblais** [comment]
Perfect thank you!

@dmeza Can you help resolve the merge conflicts?

I've also requested one dev review from Chris given code changes since the latest dev review (Joram is away for another day or so)

**Comment by dmeza** [comment]
@jasonblais @crspeller rebased off the latest from master and resolved conflicts. Ready for review.

**Comment by crspeller** [comment]
@dmeza Just needs another rebase and we can merge.

**Comment by dmeza** [comment]
@crspeller rebased off the latest from master. All checks passed. Thank you.

**Pull request merged**